### PR TITLE
changes aiding makefile migration to python execution

### PR DESCRIPTION
**Pull request by cpelley**
- Adds a UI accesible from paraflow (`command_executor`).
- @TomekTrzeciak's `tee` argument which enables saving to disk AND passing through the output (not used for Python execution right now but def. will come in handy).  Argument now `pass_through_output`.
- New `TimeIt` class for printing execution elapsed time when running with verbose.

## Background

A variation of the implementation represented in https://github.com/MetOffice/improver_suite/pull/1203

## Issues

https://github.com/MetOffice/improver_suite/issues/1141

## Dependencies

Dependency of https://github.com/MetOffice/improver_suite/pull/1218

### CODE REVIEW

**Comment by cpelley** [comment]
Ready for review.  Note that further work will follow the merge of this PR (incremental changes).

File: improver/cli/__init__.py
Comment (BelligerG) [inline improver/cli/__init__.py:343]: I like having this option, but I'm not a fan of the name :-)
It's in reference to bash `tee`? My preference would be to call it something more transparent (but it is personal preference, so up to you)

Comment (LaurenceBeard) [reply improver/cli/__init__.py:343]: I must confess I find this a useful description - as it mirrors what shell tee does.. any thoughts for a clearer alternative?

File: improver/cli/__init__.py
Comment (LaurenceBeard) [inline improver/cli/__init__.py:485]: No circumstance you'd want to use query during execution?

File: improver/cli/__init__.py
Comment (LaurenceBeard) [inline improver/cli/__init__.py:345]: This will a useful feature to have (whether or not it could be done asynchronously)

File: improver/cli/__init__.py
Comment (BelligerG) [reply improver/cli/__init__.py:343]: Because you know shell ;-)
But a lot of the devs for improver don't, which means this won't be as obvious to them, how about  something really clear, e.g. `pass_through_output`. 
Actually, being really picky, this doesn't behave like shell `tee`, if we specify `output` as `False` then `tee` actually does nothing

Comment (cpelley) [reply improver/cli/__init__.py:343]: I didn't originally know about the tee bash reference myself until I looked it up.
`pass_through_output` has my vote 👍

Comment (cpelley) [reply improver/cli/__init__.py:343]: Or perhaps `always_pass_through_output` might be more descriptive?

File: improver/cli/__init__.py
Comment (cpelley) [reply improver/cli/__init__.py:485]: Sorry @LaurenceBeard, can you clarify what you mean?
I thought it useful to provide access to the elapsed time as a numeric value (perhaps when it comes to logging).
Are you suggesting removing this property? (we don't use it right now)

Comment (LaurenceBeard) [reply improver/cli/__init__.py:485]: Nope, just pointing out that it'll be None until exit, was checking there was no case for this (it appears not!)

File: improver/cli/__init__.py
Comment (LaurenceBeard) [reply improver/cli/__init__.py:343]: yeah, any variation of pass through, or return output 👍 
would it want to be the default? we usually want to return the output even if writing out?

Comment (cpelley) [reply improver/cli/__init__.py:343]: > would it want to be the default? we usually want to return the output even if writing out?

I think that might make sense given that I *think* it would only be the end of the chains that wouldn't need passing their results on in memory.

Comment (TomekTrzeciak) [reply improver/cli/__init__.py:343]: > > would it want to be the default? we usually want to return the output even if writing out?
> 
> I think that might make sense given that I _think_ it would only be the end of the chains that wouldn't need passing their results on in memory.

FYI, clize will by default print the return value if it's not `None`, which is why I haven't made it the default.

**Comment by cpelley** [comment]
I have rebased and pointing this PR to master.

**Review by LaurenceBeard — APPROVED** [review]

File: improver/cli/__init__.py
Comment (BelligerG) [inline improver/cli/__init__.py:510]: We need a logging approach in improver too.. I wonder if we should set up something basic, then it might make a nice first task going through a few CLIs and adding some logging calls?

**Review by BelligerG — COMMENTED** [review]
This is looking really good, I think it's safe to merge, it shouldn't impact current improver usage

**Comment by cpelley** [comment]
You happy to approve @LaurenceBeard, @BelligerG?

**Comment by cpelley** [comment]
> This is looking really good, I think it's safe to merge, it shouldn't impact current improver usage

@BelligerG can you confirm by clicking "approve"? :)

> We need a logging approach in improver too.. I wonder if we should set up something basic, then it might make a nice first task going through a few CLIs and adding some logging calls?

Let me know if you intended for me to setup a simple logger now.  I wasn't sure whether you wanted a logger as part of a review action.

Cheers

**Comment by BelligerG** [comment]
> > This is looking really good, I think it's safe to merge, it shouldn't impact current improver usage
> @BelligerG can you confirm by clicking "approve"? :)

I can :-)

> Let me know if you intended for me to setup a simple logger now. I wasn't sure whether you wanted a logger as part of a review action.

Oh no, I think it'd be a massive (tedious) task and out of the scope of this PR.

**Review by BelligerG — DISMISSED** [review]
Looks good to me!

**Review by LaurenceBeard — APPROVED** [review]
Happy for this to go in when you're ready

**Comment by cpelley** [comment]
@LaurenceBeard you happy to re-approve? - I have since rebased since.
I think I'll just merge this straight in after approval since I don't see that changes in improver are likely now.

**Comment by cpelley** [comment]
@gavinevans correctly points out that I have neglected to give context to the change in behaviour of the `map_to_files` CLI.
This change relates to the fact that sometimes the `local_time` argument is provided by some bash commands to parse the output filepath.  See here:
https://github.com/MetOffice/improver_suite/blob/master/lib/python/mo/settings/app/map_to_timezones.py#L9-L10

To avoid mixing shell and python in the call, we adopt moving this parsing to improver.
For now we adopt a minimal change to the CLI (backwards compatible) which allows us to pass the output filepath as `local_time` (the application will then parse the time from this filepath).

Medium-long term, it would be desirable to make the `local_time` argument optional.  When not provided, `map_to_files` would determined the time by parsing the filepath provided by the `--output` argument.  That is, we would no longer we providing $OUTPUT to `local_time`.
Since this requires a more radical change in CLI, I held off from making this change in the first instance.

**Comment by cpelley** [comment]
Made a minor change to pull the import into process on request from @gavinevans 👍

**Review by LaurenceBeard — APPROVED** [review]

**Pull request merged**
